types/hooks.d.ts

```diff
@@ -0,0 +1,235 @@
+    /**
+     * Values that apply to elements that are neither entering nor leaving (you can use this to update present elements), or: item => values
+     */
+    update?: InferFrom<DS> | ((item: TItem) => InferFrom<DS>)
+  }
+
+  type TransitionState = 'enter' | 'leave' | 'update'
+
+  interface UseTransitionResult<TItem, DS extends object> {
+    item: TItem
+    key: string
+    state: TransitionState
+    props: ForwardedProps<DS>
+  }
+
+  export function useTransition<TItem, DS extends object>(
+    values: Merge<DS, UseTransitionProps<TItem, DS>>
+  ): UseTransitionResult<TItem, ForwardedProps<DS>>[] // result array is safe to modify
+
+  // higher order hooks 🤯
+  export namespace useKeyframes {
+    // this kind of higher order hook requires higher kinded types to type correctly at all
+    // this is the best approximation I can get to....... and there sure are anys around 🤯
+
+    // the docs says it receives a 3rd argument with the component's own props, but it's only
+    // ever called with two (https://github.com/react-spring/react-spring/blob/v7.1.3/src/hooks/KeyframesHook.js#L25)
+    export type KeyframeFn<DS extends object> = (
+      next: (props: UseSpringProps<DS>) => Promise<void>,
+      cancel: () => void
+    ) => Promise<void>
+    interface SpringKeyframeSlotsConfig extends HooksBaseProps {
+      // this is way too self-referential to type correctly
+      // both onRest and config have to have vague types... 😭
+
+      /**
+       * Callback when the animation comes to a still-stand
+       */
+      onRest?(ds: any): void
+      config?:
+        | SpringConfig
+        | ReadonlyArray<SpringConfig>
+        | ((slot: string) => SpringConfig | ReadonlyArray<SpringConfig>)
+    }
+    interface TrailKeyframeSlotsConfig<TItem>
+      extends SpringKeyframeSlotsConfig {
+      items: ReadonlyArray<TItem>
+    }
+
+    type ResolveKeyframeSlotValue<T> = T extends KeyframeFn<infer V>
+      ? V
+      : T extends ReadonlyArray<infer U>
+      ? U extends KeyframeFn<infer V>
+        ? V
+        : U extends Function // guard against accidentally putting in a KeyframeFn function
+        ? never
+        : U
+      : T extends Function // same guard, but when the function is not in an array
+      ? never
+      : T
+
+    // fun fact: the state is literally named "undefined" if you're using this overload
+    // the docs are vague but it seems this just loops the one animation forever?
+    export function spring<DS extends object>(
+      animation:
+        | ReadonlyArray<DS>
+        | ReadonlyArray<KeyframeFn<DS>>
+        | KeyframeFn<DS>,
+      initialProps?: UseSpringProps<DS>
+    ): UseSpringKeyframes<DS>
+    // unfortunately, it's not possible to infer the type of the callback functions (if any are given)
+    // while also remaining possible to infer the slot names. Callback functions have to be cast with
+    // `as useKeyframes.KeyframeFn<{ ... }>`.
```

Comment:
i.e. you have to write

```ts
spring({
  foo: (async next => next({ x: 0 })) as KeyframeFn<{ x: number }>
})
```

because of how we cannot tell TypeScript the possible function type of an arbitrary key.

If we add an index type, it then is no longer possible to have `config` be a function that receives a string, or if we _do_ make it possible, then there are two possible function signatures to pick so TypeScript picks neither and we're back at the implicit any problem.

_Even then_, because of how TypeScript changes its inference when calculating a generic argument, it would not work. TypeScript would first derive an entirely new type for the argument (similar to doing `const x = (the argument)`), and _then_ check if the derived type fits the `extends` clause. It's similar to subclassing -- subclasses in TypeScript don't inherit their parents' types when they override methods, instead they get their very own fresh method type, and then later the compiler checks if it is actually compatible with the superclass'.

Making it not generic so we could use the index signature goes back to problem 1, `config` can also be a function and it has a different signature, but its signature has to be in the index type too.

Reply:
An alternative I thought of is to also have the _keys_ be generic; this would allow us to not have to use an index signature.

```ts
    export function spring<TSlot extends string, DS extends object>(
      slots: Record<TSlot, DS | ReadonlyArray<DS> | KeyframeFn<DS>> & SpringKeyframeSlotsConfig,
      // also unfortunately not possible to strongly type this either
      initialProps?: UseSpringProps<any>
    ): UseSpringKeyframesWithSlots<Record<TSlot, DS>>
```

However, because there's no way to indicate to typescript that something that extends `object` must also not be a function, TS just infers `DS` as a function signature with an implicit any and it's back to problem 1 again.

Reply:
I wonder if I'm not doing things backwards; maybe what you _do_ need is to pass your component type (or props) as a generic so we can derive what to Pick from it.

But I still can't find a good DX for it, not until TypeScript has placeholder types (should be in 3.3?)